File: problems_on_please/CpyAug2012/contests/day06/brackets/solutions/solution.cpp
#include<iostream>
#include <cstdio>

using namespace std;

void Gen(int n, int k, string Prefix)
{
	if(n==0)
		cout<<Prefix<<endl;
	else
	{
		if(k+1<=n-1)
			Gen(n-1,k+1,Prefix+"(");
		if(k>0)
			Gen(n-1,k-1,Prefix+")");
	}
}

int main()
{
    freopen("brackets.in", "r", stdin);
    freopen("brackets.out", "w", stdout);
    int n;
	cin>>n;
	Gen(2*n,0,"");
	return 0;
}

File: problems_on_please/CpyAug2012/contests/day06/brackets/solutions/solution.py
#!/usr/bin/python3
from sys import *
setrecursionlimit(100000)

fin = open('brackets.in', 'r')
fout = open('brackets.out', 'w')

def gen(prefix, n, cnt_open, cnt_close):
    if len(prefix) == 2 * n:
        print(prefix, file=fout)
    else:
        if n - cnt_open > 0:
            gen(prefix + '(', n, cnt_open + 1, cnt_close)
        if cnt_open - cnt_close > 0:
            gen(prefix + ')', n, cnt_open, cnt_close + 1)

n = int(fin.read().strip())
gen('', n, 0, 0)

fout.close()
